- added 2 new models - MapRating, Genre - added DB structure: schema and initial inputs - added H2 DB connection - added dto, mappers to table, db storage impl - film and user services were updated to db storage logic - added 2 new controllers to cover /genre and /mpa request - added functionality to confirm friendship

imposmas opened the pull request.

imposmas committed:
- added 2 new models - MapRating, Genre
- added DB structure: schema and initial inputs
- added H2 DB connection
- added dto, mappers to table, db storage impl
- film and user services were updated to db storage logic
- added 2 new controllers to cover /genre and /mpa request
- added functionality to confirm friendship

imposmas committed:
- checkstyle correction

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/mapper/FilmMapper.java:
```diff
@@ -0,0 +1,97 @@
+        dto.setDuration(film.getDuration());
+
+        if (film.getMpaRating() != null) {
+            MpaRating mpaRating = mpaStorage.findById(film.getMpaRating())
```
Здесь будет доп. запрос к БД. Можно его избежать, если в "SELECT * FROM FILMS" заджойнить mpa и здесь его просто создать через new 

ps 
Вообще доп запросы в маппере считается не очень хорошей практикой, т.к. маппер должен просто преобразовать один объект в другой. А логику соединения разных объектов лучше вынести в классы сервисы.

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/mapper/FilmMapper.java:
```diff
@@ -0,0 +1,97 @@
+
+        if (film.getGenres() != null && !film.getGenres().isEmpty()) {
+            Set<Genre> genres = film.getGenres().stream()
+                    .map(id -> genreStorage.findById(id)
```
Здесь будет цикл с запросами к БД. Оптимальней получить все жанры для фильма одним запросом с помощью select ... in https://tproger.ru/articles/sql-in

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmStorageImpl.java:
```diff
@@ -1,8 +1,10 @@
 
+@Deprecated
 @Repository
 public class FilmStorageImpl extends InMemoryStorage<Film> implements FilmStorage {
```
InMemory реализации уже можно удалить, больше не понадобятся

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/exception/ExceptionControllerAdvice.java:
```diff
@@ -13,31 +15,26 @@
 @RestControllerAdvice
 public class ExceptionControllerAdvice {
```
Здесь все отлично

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/dal/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,153 @@
+        if (film.getGenres() == null || film.getGenres().isEmpty()) return;
+
+        String sql = "INSERT INTO FILM_GENRES (FILM_ID, GENRE_ID) VALUES (?, ?)";
+        film.getGenres().forEach(genreId -> jdbcTemplate.update(sql, film.getId(), genreId));
```
В цикле вставлять по одной записи не эффективно. Лучше воспользоваться пакетной операцией https://habr.com/ru/articles/703828/

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/dal/storage/AbstractDbStorage.java:
```diff
@@ -0,0 +1,17 @@
+import java.util.Collection;
+import java.util.Optional;
+
+public abstract class AbstractDbStorage<T> implements Storage<T> {
```
В чем смысл абстрактного класса который просто копирует интерфейс? Вот если бы он реализовывал базовые операции, тогда имело бы смысл.

VadimZharkov commented on src/main/resources/schema.sql:
Хорошая схема данных

VadimZharkov reviewed (changes requested):
Хорошая работа, посмотри, что можно улучшить.

imposmas committed:
- removed DB requests from mapper
- deleted InMemory classes
- changed insert cycle to batchUpdate
- moved basic methods to AbstractDbStorage

imposmas committed:
- checkstyle
